File: Classes/ScoreCard.m
//
//  ScoreCard.m
//  MathFlashCards
//
//  Created by Jagmeet Chawla on 4/12/09.
//  Revised by Krishna Narayan on 9/17/09
//		--Updated graphics
//		--Added tell a friend
//		--Changed app name to Kids Learn To Tell Time
//  Copyright 2009-2013 NSC Partners LLC. All rights reserved.
//

#import "ScoreCard.h"

@implementation ScoreCard

@synthesize playerName;
@synthesize activity;
@synthesize activityType;
@synthesize activityLevel;
@synthesize questionsAsked;
@synthesize questionsAttempted;
@synthesize rightAnswers;
@synthesize wrongAnswers;
@synthesize percentScore;
@synthesize secondsTaken;
@synthesize scoreDateTime;
@synthesize scoreCard;
@synthesize scoreRank;
@synthesize isTopScore;

- (NSDictionary *) newScoreCard {
	NSMutableDictionary *newScore = [[NSMutableDictionary alloc] init];
	[newScore setObject:[NSString stringWithString:self.playerName] forKey:kPlayerName];
	[newScore setObject:[NSNumber numberWithInt:self.activity] forKey:kActivity];
	[newScore setObject:[NSNumber numberWithInt:self.activityType] forKey:kActivityType];
	[newScore setObject:[NSNumber numberWithInt:self.activityLevel] forKey:kActivityLevel];
	[newScore setObject:[NSNumber numberWithInt:self.questionsAsked] forKey:kQuestionsAsked];
	[newScore setObject:[NSNumber numberWithInt:self.questionsAttempted] forKey:kQuestionsAttempted];
	[newScore setObject:[NSNumber numberWithInt:self.rightAnswers] forKey:kRightAnswers];
	[newScore setObject:[NSNumber numberWithInt:self.wrongAnswers] forKey:kWrongAnswers];
	[newScore setObject:[NSNumber numberWithFloat:self.percentScore] forKey:kPercentScore];
	[newScore setObject:[NSNumber numberWithInt:self.secondsTaken] forKey:kSecondsTaken];
	[newScore setObject:[NSDate date] forKey:kScoreDateTime];
	
	//pre-load topscore boolean and rank variables
	isTopScore = NO;
	scoreRank = -1; //-1 denotes unranked
		
	if (rightAnswers > 0) {
		//Now get the file name - root for documents directory + "/Scores/"
		NSString *docPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
		NSString *filePath = [docPath stringByAppendingString:@"/"];
		NSString *fileName = [NSString stringWithFormat:kFileVarScores,self.activity,self.activityType,self.activityLevel];
		NSString *fileNameWithPath = [filePath stringByAppendingString:fileName];
		//First read scores array from disk
		NSArray *scoresArray = [[NSArray alloc] initWithContentsOfFile:fileNameWithPath];
		
		//Now Loop throgh the sorted array and see if the score is greater than any other score in the array
		if ([scoresArray count] <= [KidsTimeFunAppState sharedState].sizeOfTopScoreList) {
			isTopScore = YES;
			scoreRank = [scoresArray count]+1;
		}
		for (int i=0; i < [scoresArray count]; i++) {
			//if first - just add is as top score #1
			if ([scoresArray count] == 0) (scoreRank = 1);
			//initialize sort keys
			long long thisSortKey = 0, arrSortKey = 0;
			//sortkey creators
			int thisNumQT, thisNumQR, thisSec, arrNumQT, arrNumQR, arrSec, thisPer10000, arrPer10000;
			switch (self.activityType) {
				case kActTypeNumbered:
					//create sort key - for numbered quiz the rank order is %, total number of questions, time
					//time should be in descending order, so subtract seconds from 15 minutes i.e. 900 seconds
					thisPer10000 = (int) (self.percentScore*10000.0f);
					thisNumQT = self.questionsAsked;
					thisSec = 900-self.secondsTaken;
					thisSortKey = [[NSString stringWithFormat:@"%i%i%i",thisPer10000,thisNumQT,thisSec] longLongValue];
					arrPer10000 = (int) ([[[scoresArray objectAtIndex:i] valueForKey:kPercentScore] floatValue]*10000.0f);
					arrNumQT = [[[scoresArray objectAtIndex:i] valueForKey:kQuestionsAsked] intValue];
					arrSec = 900-[[[scoresArray objectAtIndex:i] valueForKey:kSecondsTaken] intValue];
					arrSortKey = [[NSString stringWithFormat:@"%i%i%i",arrPer10000,arrNumQT,arrSec] longLongValue];					
					break;
				case kActTypeTimed:
					//create sort key - for timed quiz the rank order is time, number of questions right, total number of questions
					thisSec = 900-self.secondsTaken;
					thisNumQT = self.questionsAsked;
					thisNumQR = self.rightAnswers;
					thisSortKey = [[NSString stringWithFormat:@"%i%i%i",thisSec,thisNumQT,thisNumQR] longLongValue];
					arrSec = 900-[[[scoresArray objectAtIndex:i] valueForKey:kSecondsTaken] intValue];
					arrNumQT = [[[scoresArray objectAtIndex:i] valueForKey:kQuestionsAsked] intValue];
					arrNumQR = [[[scoresArray objectAtIndex:i] valueForKey:kRightAnswers] intValue];
					arrSortKey = [[NSString stringWithFormat:@"%i%i%i",arrSec,arrNumQT,arrNumQR] longLongValue];
					break;
				default:
					break;
			}
			//now the moment thissortkey is greater than or equal the array sort key - that is the rank
			//it assumes the data is in sorted order
			//which is the responsibility of file writer to write it in this sorted order
			if (thisSortKey >= arrSortKey) {
				scoreRank = i+1;
				isTopScore = YES;
				break; //break out of the loop the moment you find a rank
			}
		} //for loop
		//now release the array
		[scoresArray release];
	} //if rightanswers > 0
	//set rank and top scores
	[newScore setObject:[NSNumber numberWithInt:self.scoreRank] forKey:kScoreRank];
	scoreCard = [[NSDictionary dictionaryWithDictionary:newScore] retain];
	[newScore release];
	return scoreCard;
}

- (BOOL) writeScoreCard {
	BOOL returnValue = NO;
	if (!self.isTopScore) return returnValue;
	//Now get the file name - root for documents directory + "/Scores/"
	NSString *docPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
	NSString *filePath = [docPath stringByAppendingString:@"/"];
	NSString *fileName = [NSString stringWithFormat:kFileVarScores,self.activity,self.activityType,self.activityLevel];
	NSString *fileNameWithPath = [filePath stringByAppendingString:fileName];
	//First read scores array from disk
	NSArray *scoresArray = [[NSArray alloc] initWithContentsOfFile:fileNameWithPath];
	//Create New Scores Array to write
	//Load elements 0 to rank-1 from old array
	//then current score
	//then rank+1 to size of array-1 from old array
	NSMutableArray *newScoresArray = [[NSMutableArray alloc] init];
	for (int i=0; i<self.scoreRank-1; i++) {
		[newScoresArray addObject:[scoresArray objectAtIndex:i]];
	}
	[newScoresArray addObject:self.scoreCard];
	int newArraySize = ([scoresArray count]<[KidsTimeFunAppState sharedState].sizeOfTopScoreList)?[scoresArray count]+1:[KidsTimeFunAppState sharedState].sizeOfTopScoreList;
	for (int i=self.scoreRank; i<newArraySize; i++) {
		[newScoresArray addObject:[scoresArray objectAtIndex:i-1]];
	}
	//Now write file
	NSString *errorDesc;
	NSData *plistData = [NSPropertyListSerialization dataFromPropertyList:newScoresArray format:NSPropertyListXMLFormat_v1_0 errorDescription:&errorDesc];
    if (plistData) {
		returnValue = [plistData writeToFile:fileNameWithPath atomically:NO];
    }
    else {
        //NSLog(errorDesc);
        [errorDesc release];
		returnValue = NO;
	}
	[scoresArray release];
	[newScoresArray release];
	return returnValue;
}

- (void) dealloc {
	[scoreDateTime release];
	[playerName release];
	[scoreCard release];
	[super dealloc];
}

@end

File: Classes/ScoreCard.swift
import Foundation

class ScoreCard: NSObject {

    var playerName: String = ""
    var activity: Int32 = 0
    var activityType: Int32 = 0
    var activityLevel: Int32 = 0
    var questionsAsked: Int32 = 0
    var questionsAttempted: Int32 = 0
    var rightAnswers: Int32 = 0
    var wrongAnswers: Int32 = 0
    var percentScore: Float = 0
    var secondsTaken: Int32 = 0
    private(set) var scoreRank: Int32 = -1
    private(set) var scoreDateTime: Date?
    private(set) var scoreCard: [String: Any]?
    private(set) var isTopScore: Bool = false

    private var scoresFilePath: String {
        let docs = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let fileName = String(format: kFileVarScores, activity, activityType, activityLevel)
        return docs + "/" + fileName
    }

    @discardableResult
    func newScoreCard() -> [String: Any] {
        var card: [String: Any] = [
            kPlayerName: playerName,
            kActivity: NSNumber(value: activity),
            kActivityType: NSNumber(value: activityType),
            kActivityLevel: NSNumber(value: activityLevel),
            kQuestionsAsked: NSNumber(value: questionsAsked),
            kQuestionsAttempted: NSNumber(value: questionsAttempted),
            kRightAnswers: NSNumber(value: rightAnswers),
            kWrongAnswers: NSNumber(value: wrongAnswers),
            kPercentScore: NSNumber(value: percentScore),
            kSecondsTaken: NSNumber(value: secondsTaken),
            kScoreDateTime: Date()
        ]

        isTopScore = false
        scoreRank = -1

        if rightAnswers > 0 {
            let existing = (NSArray(contentsOfFile: scoresFilePath) as? [[String: Any]]) ?? []

            if existing.count <= KidsTimeFunAppState.sharedState().sizeOfTopScoreList {
                isTopScore = true
                scoreRank = Int32(existing.count) + 1
            }

            for (i, entry) in existing.enumerated() {
                let thisSortKey = sortKey(for: self)
                let arrSortKey = sortKey(fromDict: entry)
                if thisSortKey >= arrSortKey {
                    scoreRank = Int32(i) + 1
                    isTopScore = true
                    break
                }
            }
        }

        card[kScoreRank] = NSNumber(value: scoreRank)
        scoreCard = card
        return card
    }

    func writeScoreCard() -> Bool {
        guard isTopScore, let card = scoreCard else { return false }
        let existing = (NSArray(contentsOfFile: scoresFilePath) as? [[String: Any]]) ?? []
        var updated: [[String: Any]] = []
        for i in 0..<(Int(scoreRank) - 1) where i < existing.count {
            updated.append(existing[i])
        }
        updated.append(card)
        let maxSize = min(existing.count + 1, Int(KidsTimeFunAppState.sharedState().sizeOfTopScoreList))
        for i in Int(scoreRank)..<maxSize where (i - 1) < existing.count {
            updated.append(existing[i - 1])
        }
        guard let data = try? PropertyListSerialization.data(fromPropertyList: updated, format: .xml, options: 0) else { return false }
        return (try? data.write(to: URL(fileURLWithPath: scoresFilePath))) != nil
    }

    private func sortKey(for card: ScoreCard) -> Int64 {
        switch card.activityType {
        case kActTypeNumbered:
            let pct = Int(card.percentScore * 10000)
            let q = Int(card.questionsAsked)
            let s = 900 - Int(card.secondsTaken)
            return Int64("\(pct)\(q)\(s)") ?? 0
        case kActTypeTimed:
            let s = 900 - Int(card.secondsTaken)
            let q = Int(card.questionsAsked)
            let r = Int(card.rightAnswers)
            return Int64("\(s)\(q)\(r)") ?? 0
        default:
            return 0
        }
    }

    private func sortKey(fromDict d: [String: Any]) -> Int64 {
        switch activityType {
        case kActTypeNumbered:
            let pct = Int(((d[kPercentScore] as? NSNumber)?.floatValue ?? 0) * 10000)
            let q = (d[kQuestionsAsked] as? NSNumber)?.intValue ?? 0
            let s = 900 - ((d[kSecondsTaken] as? NSNumber)?.intValue ?? 0)
            return Int64("\(pct)\(q)\(s)") ?? 0
        case kActTypeTimed:
            let s = 900 - ((d[kSecondsTaken] as? NSNumber)?.intValue ?? 0)
            let q = (d[kQuestionsAsked] as? NSNumber)?.intValue ?? 0
            let r = (d[kRightAnswers] as? NSNumber)?.intValue ?? 0
            return Int64("\(s)\(q)\(r)") ?? 0
        default:
            return 0
        }
    }
}
